--- yp-tools-2.14/src/yppasswd.c.adjunct	2017-01-04 08:40:17.081225336 +0100
+++ yp-tools-2.14/src/yppasswd.c	2017-01-04 08:41:01.214077346 +0100
@@ -450,6 +450,7 @@
   passwdlen = get_passwd_len (pwd->pw_passwd);
   if (pwd->pw_passwd[0] 
       && 0 != strcmp (pwd->pw_passwd, "x") /* don't check shadow passwords */
+      && 0 != strncmp (pwd->pw_passwd, "##", 2) /* don't check passwords using passwd.adjunct feature */
       && uid)
     {
       char *crypted = crypt(pwdstr, pwd->pw_passwd);
@@ -774,6 +775,7 @@
       /* We can't check the password with shadow passwords enabled. We
        * leave the checking to yppasswdd */
       if (uid != 0 && strcmp (pwd->pw_passwd, "x") != 0 &&
+	  0 != strncmp (pwd->pw_passwd, "##", 2) && /* don't check passwords using passwd.adjunct feature */
 	  strcmp (pwd->pw_passwd, hashpass ) != 0)
 	{
 	  int passwdlen = get_passwd_len (pwd->pw_passwd);
